runtime/router/trie.go

```diff
@@ -0,0 +1,293 @@
+		i++
+	}
+
+	// find index j, k in key and path to which they match,
+	// j and k is only useful to check if there is conflict,
+	// they are not where splits happen, splits happen at index i.
+	var j, k int
+	for j < m && k < n {
+		if t.key[j] == ':' || path[k] == ':' {
+			oj, ok := j, k
+			same := t.key[j] == path[k]
+			for j < m && t.key[j] != '/' {
+				j++
+			}
+			for k < n && path[k] != '/' {
+				k++
+			}
+			if same && (j-oj) != (k-ok) {
+				return &paramMismatch{
+					t.key[oj:j],
+					path[ok:k],
+					t.key,
+				}
+			}
+		} else if t.key[j] == path[k] {
+			j++
+			k++
+		} else {
+			break
+		}
+	}
+
+	// conflicts caused by ":" is only possible when j == m
+	if j == m {
+		for _, c := range t.children {
+			if c.key[0] == path[k] || c.key[0] == ':' || path[k] == ':' {
+				if _, _, err := c.get(path[k:]); err == nil {
+					return errExist
+				}
+			}
+		}
+	}
+
+	// node ley is longer than longest common prefix
+	if i < m {
+		// key/path suffix being "*" means a conflict
+		if path[i:] == "*" || t.key[i:] == "*" {
+			return errExist
+		}
+
+		// split the node key, add new node with node key minus longest common prefix
+		split := &tnode{
+			key:      t.key[i:],
+			value:    t.value,
+			children: t.children,
+		}
+		t.key = t.key[:i]
+		t.value = nil
+		t.children = []*tnode{split}
+
+		// path is equal to longest common prefix
+		// set value on current node after split
+		if i == n {
+			t.value = value
+		} else {
+			// path is longer than longest common prefix
+			// add new node with path minus longest common prefix
+			newNode := &tnode{
+				key:   path[i:],
+				value: value,
+			}
+			t.children = append(t.children, newNode)
+		}
+	}
+
+	// node key is equal to longest common prefix
+	if i == m {
+		// path is equal to longest common prefix
+		if i == n {
+			// node is guaranteed to have zero value,
+			// otherwise it would have caused errExist earlier
+			t.value = value
+		} else {
+			// path is longer than node key, try to recurse on node children
+			for _, c := range t.children {
+				if c.key[0] == path[i] {
+					err := c.set(path[i:], value)
+					if e, ok := err.(*paramMismatch); ok {
+						e.existingPath = t.key + e.existingPath
+						return e
+					}
+					return err
+				}
+			}
+			// no children to recurse, add node with path minus longest common path
+			newNode := &tnode{
+				key:   path[i:],
+				value: value,
+			}
+			t.children = append(t.children, newNode)
+		}
+	}
+
+	return nil
+}
+
+func (t *tnode) get(path string) (http.Handler, []Param, error) {
+	m, n := len(t.key), len(path)
+	var params []Param
+
+	// find the longest matched prefix
+	var j, k int
+	for j < m && k < n {
+		if t.key[j] == ':' {
+			oj, ok := j+1, k
+			for j < m && t.key[j] != '/' {
+				j++
+			}
+			for k < n && path[k] != '/' {
+				k++
+			}
+			params = append(params, Param{t.key[oj:j], path[ok:k]})
+		} else if path[k] == ':' { // necessary for conflict check used in set call
```

Comment:
Maybe is unexpected that this would also be used in regular request path, so that this is a successful test:

```
func TestTrieMatch(t *testing.T) {
	tree := NewTrie()
	tests := []ts{
		{op: set, path: "/a", value: "foo"},
		{op: get, path: "/:a", expectedValue: "foo"},
```

Reply:
yeah, I will pass in a flag to indicate whether to treat `:` as a pattern or not in query path.

Reply:
Introduced two new flags to keep track of whether the current colon is a pattern, now this case returns not found.